scheds/rust/scx_rusty/src/load_balance.rs

```diff
@@ -0,0 +1,1082 @@
+        // counterpart while scanning right and picking the better of the
+        // two.
+        let tasks = std::mem::take(&mut push_dom.tasks).into_vec();
+        let (task, new_imbal) = match (
+            Self::find_first_candidate(
+                tasks
+                    .as_slice()
+                    .iter()
+                    .filter(|x| x.load <= OrderedFloat(to_xfer))
+                    .rev(),
+                pull_dom.id.try_into().unwrap(),
+                self.skip_kworkers,
+            ),
+            Self::find_first_candidate(
+                tasks
+                    .as_slice()
+                    .iter()
+                    .filter(|x| x.load >= OrderedFloat(to_xfer)),
+                pull_dom.id.try_into().unwrap(),
+                self.skip_kworkers,
+            ),
+        ) {
+            (None, None) => return Ok(None),
+            (Some(task), None) | (None, Some(task)) => {
+                (task, calc_new_imbal(*task.load))
+            }
+            (Some(task0), Some(task1)) => {
+                let (new_imbal0, new_imbal1) = (calc_new_imbal(*task0.load), calc_new_imbal(*task1.load));
+                if new_imbal0 <= new_imbal1 {
+                    (task0, new_imbal0)
+                } else {
+                    (task1, new_imbal1)
+                }
+            }
+        };
+
+        // If the best candidate can't reduce the imbalance, there's nothing
+        // to do for this pair.
+        let old_imbal = to_push + to_pull;
+        if old_imbal < new_imbal {
+            std::mem::swap(&mut push_dom.tasks, &mut SortedVec::from_unsorted(tasks));
+            return Ok(None);
+        }
+
+        let load = *(task.load);
+        let pid = task.pid;
+        task.migrated.set(true);
+        std::mem::swap(&mut push_dom.tasks, &mut SortedVec::from_unsorted(tasks));
+
+        push_dom.transfer_load(load, pid, pull_dom, &mut self.skel);
+        Ok(Some(load))
+    }
+
+    fn transfer_between_nodes(&mut self,
+                              push_node: &mut NumaNode,
+                              pull_node: &mut NumaNode) -> Result<f64> {
+        let n_push_doms = push_node.push_domains.len();
+        let n_pull_doms = pull_node.pull_domains.len();
+        debug!("Inter node {} -> {} started ({} push domains -> {} pull domains)",
+               push_node.id, pull_node.id, n_push_doms, n_pull_doms);
+
+        if n_push_doms == 0 || n_pull_doms == 0 {
+            return Ok(0.0f64);
+        }
+
+        let push_imbal = push_node.load.imbal();
+        let pull_imbal = pull_node.load.imbal();
+        let xfer = push_node.xfer_between(&pull_node);
+
+        let mut delta = 0.0f64;
+        let mut push_doms = std::mem::take(&mut push_node.push_domains).into_vec();
+        for push_dom in push_doms.iter_mut().rev() {
+            let mut pull_doms = std::mem::take(&mut pull_node.pull_domains).into_vec();
+            for pull_dom in pull_doms.iter_mut() {
```

Comment:
Does this mean that the amount that can be balanced across nodes is limited by the existing internal imbalance in the participating nodes? If so, this may be okay in practice as there usually would be on-going perturbance in the system but there can be pathological cases - e.g. the same number of full duty-cycle high weight threads per CPU on the source or target node. Given that the imbalance is between whole nodes, the right thing to do probably is to transfer from the highest load domain to the lowest one?

Reply:
>Does this mean that the amount that can be balanced across nodes is limited by the existing internal imbalance in the participating nodes?

Sorry, I'm not sure I understand the question. Imbalance at the node level is calculated according to the per-node average. The per-domain average doesn't come into play here beyond us only being able to transfer tasks if they help equalize the imbalance, just like with cross-domain.

>Given that the imbalance is between whole nodes, the right thing to do probably is to transfer from the highest load domain to the lowest one?

That's the idea of iterating over the ordered push domains and pull domains. Or are you suggesting we just always iterate over highest load domain -> lowest load domain instead of doing two passes where we do NUMA <-> NUMA, followed by intra-NUMA (and applying a different cost for going cross-NUMA vs. just cross-domain)?

The issue with doing that is the imbalance could be solved by migrating between domains in the same node. We don't want to push to domains whose node-wide load is not imbalanced with respect to the global node load average.

Reply:
Spoke offline, I see what you're saying. We could have for example two nodes with load imbalance between them, but each having domains with exactly equal load. It's not clear whether we'd correctly partition the domains into push or pull lists in such a scenario.

Let me rejigger this to just have a node own a flat list of domains, rather than a tuple of pushers, pullers, and balancers.